File: 18.SectionsCustomModelObjects/Sample Code/SectionsCustomModelObjc/SectionsCustomModelObjc/ViewController.m
#import "ViewController.h"
#import "Models.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) IBOutlet UITableView *tableView;
@property (nonatomic, strong) NSMutableArray *contents;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self populateData];
}

- (void)populateData
{
    SectionItem *fruitSection = [[SectionItem alloc] init];
    [fruitSection setSectionName:@"Fruits"];
    
    [[fruitSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Apple" color:[UIColor redColor]]];
    [[fruitSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Banana" color:[UIColor yellowColor]]];
    [[fruitSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Grape" color:[UIColor purpleColor]]];
    
    SectionItem *veggieSection = [[SectionItem alloc] init];
    [veggieSection setSectionName:@"Vegetables"];
    
    [[veggieSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Carrot" color:[UIColor orangeColor]]];
    [[veggieSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Celery" color:[UIColor greenColor]]];
    
    SectionItem *sweetSection = [[SectionItem alloc] init];
    [sweetSection setSectionName:@"Sweets"];
    [sweetSection setSectionFooter:@"I love dessert!"];
    
    [[sweetSection sectionContents] addObject:[[FoodItem alloc] initWithName:@"Chocolate" color:nil]];

    [self setContents:[@[ fruitSection, veggieSection, sweetSection ] mutableCopy]];
}

- (FoodItem *)itemAtIndexPath:(NSIndexPath *)indexPath
{
    SectionItem *sectionItem = [[self contents] objectAtIndex:[indexPath section]];
    
    return [[sectionItem sectionContents] objectAtIndex:[indexPath row]];
}

#pragma mark - UITableViewDataSource Methods

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return [[self contents] count];
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    SectionItem *sectionItem = [[self contents] objectAtIndex:section];
    
    return [[sectionItem sectionContents] count];
}

- (NSString *)tableView:(UITableView *)tableView
titleForHeaderInSection:(NSInteger)section
{
    SectionItem *sectionItem = [[self contents] objectAtIndex:section];

    return [sectionItem sectionName];
}

- (NSString *)tableView:(UITableView *)tableView
titleForFooterInSection:(NSInteger)section
{
    SectionItem *sectionItem = [[self contents] objectAtIndex:section];
    
    return [sectionItem sectionFooter];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *identifier = @"Plain";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    
    FoodItem *foodItem = [self itemAtIndexPath:indexPath];
    
    [[cell textLabel] setText:[foodItem name]];
    
    if ([foodItem color] == nil)
    {
        [[cell textLabel] setTextColor:[UIColor blackColor]];
    }
    else
    {
        [[cell textLabel] setTextColor:[foodItem color]];
    }
    
    return cell;
}

@end

File: 18.SectionsCustomModelObjects/Sample Code/SectionsCustomModelSwift/SectionsCustomModelSwift/ViewController.swift
//
//  ViewController.swift
//  SectionsCustomModelSwift
//
//  Created by Brian Slick on 1/21/16.
//  Copyright © 2016 Brian Slick. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate
{
    @IBOutlet var tableView: UITableView!
    var contents = [SectionItem]()
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        populateData()
    }

    func populateData()
    {
        let fruitSection = SectionItem()
        fruitSection.sectionName = "Fruits"

        fruitSection.sectionContents.append(FoodItem.init(name: "Apple", color: UIColor.redColor()))
        fruitSection.sectionContents.append(FoodItem.init(name: "Banana", color: UIColor.yellowColor()))
        fruitSection.sectionContents.append(FoodItem.init(name: "Grape", color: UIColor.purpleColor()))

        let veggieSection = SectionItem()
        veggieSection.sectionName = "Vegetables"
        
        veggieSection.sectionContents.append(FoodItem.init(name: "Carrot", color: UIColor.orangeColor()))
        veggieSection.sectionContents.append(FoodItem.init(name: "Celery", color: UIColor.greenColor()))
        
        let sweetSection = SectionItem()
        sweetSection.sectionName = "Sweets"
        sweetSection.sectionFooter = "I love dessert!"
        
        sweetSection.sectionContents.append(FoodItem.init(name: "Chocolate", color: nil))
        
        contents = [ fruitSection, veggieSection, sweetSection ]
    }
    
    func itemAtIndexPath(indexPath: NSIndexPath) -> FoodItem
    {
        let sectionItem = contents[indexPath.section]
        
        return sectionItem.sectionContents[indexPath.row]
    }
    
    // MARK: - UITableViewDataSource Methods
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int
    {
        return contents.count
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        let sectionItem = contents[section]
        
        return sectionItem.sectionContents.count
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String?
    {
        let sectionItem = contents[section]
        
        return sectionItem.sectionName
    }
    
    func tableView(tableView: UITableView, titleForFooterInSection section: Int) -> String?
    {
        let sectionItem = contents[section]
        
        return sectionItem.sectionFooter
    }

    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        let identifier = "Default"
        
        var optionalCell = tableView.dequeueReusableCellWithIdentifier(identifier)
        if (optionalCell == nil)
        {
            optionalCell = UITableViewCell.init(style: .Default, reuseIdentifier: identifier)
        }
        
        let cell = optionalCell!
        
        let foodItem = itemAtIndexPath(indexPath)
        
        cell.textLabel?.text = foodItem.name
        
        if let color = foodItem.color
        {
            cell.textLabel?.textColor = color
        }
        else
        {
            cell.textLabel?.textColor = UIColor.blackColor()
        }

        return cell
    }
}

